localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py

```diff
@@ -57,20 +62,88 @@ def get_queue_attributes(self) -> dict:
+
+    def collect_messages(self, max_batch_size=10, max_batch_window=0, **kwargs) -> list[dict]:
+        # The number of ReceiveMessage requests we expect to be made in order to fill up the max_batch_size.
+        _total_expected_requests = (
+            max_batch_size + DEFAULT_MAX_RECEIVE_COUNT - 1
+        ) // DEFAULT_MAX_RECEIVE_COUNT
+
+        # The maximum duration a ReceiveMessage call should take, given how many requests
+        # we are going to make to fill the batch and the maximum batching window.
+        _maximum_duration_per_request = max_batch_window / _total_expected_requests
+
+        # Number of messages we want to receive per ReceiveMessage operation.
+        messages_per_receive = min(DEFAULT_MAX_RECEIVE_COUNT, max_batch_size)
+
+        def receive_message(num_messages: int = messages_per_receive):
+            start_request_t = time.monotonic()
+            response = self.source_client.receive_message(
+                QueueUrl=self.queue_url,
+                MaxNumberOfMessages=num_messages,
+                MessageAttributeNames=["All"],
+                MessageSystemAttributeNames=[MessageSystemAttributeName.All],
+            )
+            return response.get("Messages", []), time.monotonic() - start_request_t
+
+        batch = []
+        start_collection_t = time.monotonic()
+        while not self._shutdown_event.is_set():
+            # Adjust request size if we're close to max_batch_size
+            if (remaining := max_batch_size - len(batch)) < messages_per_receive:
+                messages_per_receive = remaining
+
+            # Return the messages received and the request duration in seconds.
+            try:
+                messages, request_duration = receive_message(messages_per_receive)
+            except Exception as e:
+                # If an exception is raised here, break the loop and return whatever
+                # has been collected early.
+                # TODO: Handle exceptions differently i.e QueueNotExist or ConnectionFailed should retry with backoff
+                LOG.warning(
+                    "Polling SQS queue failed: %s",
+                    e,
+                    exc_info=LOG.isEnabledFor(logging.DEBUG),
+                )
+                break
+
+            if messages:
+                batch.extend(messages)
+
+            time_elapsed = time.monotonic() - start_collection_t
+            if time_elapsed >= max_batch_window or len(batch) >= max_batch_size:
+                return batch
+
+            # Simple adaptive interval technique to randomly backoff between last request duration
+            # and max allowed time per request.
+            # Note: This approach assumes that a larger batching window means a user is content
+            # with waiting longer for a batch response.
+            adaptive_wait_time = random.uniform(request_duration, _maximum_duration_per_request)
+            self._shutdown_event.wait(adaptive_wait_time)
```

Comment:
This assumes the following:
- The duration of the most recent `ReceiveMessage` call indicates the current performance of the system.
- Adding backoff between `ReceiveMessage` calls, when collecting batch sizes of greater than 10, will reduce load on the gateway and and internal SQS queue.
- A larger batch window means a user is content with waiting longer for a response (value of `_maximum_duration_per_request` will be larger). For example, if a user sets:
  - `max_batch_window=300s` 
  - `max_batch_size=25`
  - This indicates a user is content with the worst case of waiting 5 minutes for their 25 records -- hence we can make the backoff per request larger.

@dfangl was this what you had in mind re: not wanting to overload the internal gateway?

Reply:
I think this is a nice approach, but could you perhaps collect some example values and their expected wait times / throughput, to see if we are on the right track with this one?

Reply:
I saw most `ReceiveMessage` requests taking around `0.0002s` when running tests against LS. Will draw up a table though